src/test/rgw/test_d4n_filter.cc

```diff
@@ -0,0 +1,2810 @@
+      req.push("HGET", bucketName + "__:null_" + TEST_OBJ + testName + "_0_0", "dirty");
+
+      response< std::string, std::string > resp;
+
+      conn->async_exec(req, resp, yield[ec]);
+
+      ASSERT_EQ((bool)ec, false);
+      EXPECT_EQ(std::get<0>(resp).value(), "0");
+      EXPECT_EQ(std::get<1>(resp).value(), "0");
+    }
+
+    // Ensure object with versioning enabled is written to backend
+    {
+      auto next = dynamic_cast<rgw::sal::FilterObject*>(objEnabled.get())->get_next();
+      EXPECT_EQ(next->load_obj_state(env->dpp, optional_yield{yield}), 0);
+      EXPECT_EQ(next->exists(), 1);
+    }
+
+    // Ensure object with versioning suspended is written to backend
+    {
+      auto next = dynamic_cast<rgw::sal::FilterObject*>(objSuspended.get())->get_next();
+      EXPECT_EQ(next->load_obj_state(env->dpp, optional_yield{yield}), 0);
+      EXPECT_EQ(next->exists(), 1);
+    }
+
+    std::string attr_val;
+    std::string oid = instance + "#0#" + std::to_string(ofs);
+    std::string location = CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + oid;
+    EXPECT_EQ(d4nFilter->get_cache_driver()->get_attr(env->dpp, location, RGW_CACHE_ATTR_DIRTY, attr_val, optional_yield({yield})), 0);
+    EXPECT_EQ(attr_val, "0");
+
+    oid = version + "#0#" + std::to_string(ofs);
+    location = CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + oid;
+    EXPECT_EQ(d4nFilter->get_cache_driver()->get_attr(env->dpp, location, RGW_CACHE_ATTR_DIRTY, attr_val, optional_yield({yield})), 0);
+    EXPECT_EQ(attr_val, "0");
+
+    {
+      boost::system::error_code ec;
+      request req;
+      req.push("FLUSHALL");
+
+      response< boost::redis::ignore_t > resp;
+
+      conn->async_exec(req, resp, yield[ec]);
+
+      ASSERT_EQ((bool)ec, false);
+    } 
+
+    conn->cancel();
+    testBucket->remove(env->dpp, true, optional_yield{yield});
+    DriverDestructor driver_destructor(static_cast<rgw::sal::D4NFilterDriver*>(driver));
+  }, rethrow);
+ 
+  io.reset();
+  io.run();
+}
+
+TEST_F(D4NFilterFixture, GetVersionedObjectWrite)
+{
+  env->cct->_conf->d4n_writecache_enabled = true;
+  env->cct->_conf->rgw_d4n_cache_cleaning_interval = 2;
+  const std::string testName = "GetVersionedObjectWrite";
+ 
+  boost::asio::spawn(io, [this, &testName] (boost::asio::yield_context yield) {
+    init_driver(yield);
+    create_bucket(testName, yield);
+    std::string instance;
+    put_version_enabled_object(testName, instance, yield);
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+    put_version_suspended_object(testName, yield);
+
+    std::unique_ptr<rgw::sal::Object::ReadOp> read_op_enabled(objEnabled->get_read_op());
+    EXPECT_EQ(read_op_enabled->prepare(optional_yield{yield}, env->dpp), 0);
+    EXPECT_EQ(read_op_enabled->iterate(env->dpp, 0, ofs, nullptr, optional_yield{yield}), 0);
```

Comment:
This test must validate the data returned against the expected data, so that we are sure that data being read from the cache is correct.